src/veins/modules/utility/TimerManager.h

```diff
@@ -0,0 +1,219 @@
+   *
+   * The specified timer will start immediately but skip its first execution. In order to include the first execution, use setRelativeStart(0).
+   *
+   * Any previously set start time will be overwritten.
+   */
+  TimerSpecification &setImmediateStart();
+
+  /**
+   * Set the timer's end time.
+   *
+   * Any previously set end time will be overwritten.
+   *
+   * @param end The relative end time. It is relative to the current simtime, i.e. passing simtime_t(1, SIMTIME_S) will execute the timer until one second has passed.
+   */
+  TimerSpecification &setRelativeEnd(omnetpp::simtime_t end);
+
+  /**
+   * Set the timer's end time.
+   *
+   * Any previously set end time will be overwritten.
+   *
+   * @param end The absolute end time. The latest possible occurence is at this time. Values before the current start time will prevent any executions.
+   */
+  TimerSpecification &setAbsoluteEnd(omnetpp::simtime_t end);
+
+  /**
+   * Set the timer to be open ended.
+   *
+   * Any previously set end time will be overwritten.
+   */
+  TimerSpecification &setOpenEnd();
+
+  /**
+   * Sets the timer to execute once in a given time.
+   */
+  TimerSpecification &setOneshotIn(omnetpp::simtime_t in);
+
+  /**
+   * Sets the timer to execute once at a given time.
+   */
+  TimerSpecification &setOneshotAt(omnetpp::simtime_t at);
+
+private:
+  friend TimerManager;
+
+  enum class StartMode { RELATIVE, ABSOLUTE, IMMEDIATE };
+  enum class EndMode { RELATIVE, ABSOLUTE, REPITITION, OPEN };
+
+  /**
+   * Finalizes this instance such that its values are independent of current simulation time.
+   *
+   * After calling this function, start_mode_ is guaranteed to be StartMode::ABSOLUTE and end_mode_ to be EndMode::ABSOLUTE or EndMode::OPEN.
+   * Cannot be called on TimerSpecifications
+   */
+  void finalize();
+
+  /**
+   */
+  bool valid() const { return period_ != -1; }
+
+  /**
+   * Check that the given time is a valid occurence for this timer.
+   */
+  bool validOccurence(omnetpp::simtime_t time) const;
+
+  StartMode start_mode_;           ///< Interpretation of start time._
+  omnetpp::simtime_t start_;       ///< Time of the Timer's first occurence. Interpretation depends on start_mode_.
+  EndMode end_mode_;               ///< Interpretation of end time._
+  unsigned end_count_;             ///< Number of repititions of the timer. Only valid when end_mode_ == REPITITION.
+  omnetpp::simtime_t end_time_;    ///< Last possible occurence of the timer. Only valid when end_mode_ != REPITITION.
+  omnetpp::simtime_t period_;      ///< Time between events.
+  std::function<void()> callback_; ///< The function to be called when the Timer is triggered.
+};
+
+class TimerManager {
+private:
+  /**
+   * A message which is used for triggering Timers.
+   *
+   * Its implementation is empty as it is only used to differentiate from other
+   * messages.
+   */
+  struct TimerMessage : public omnetpp::cMessage {
```

Comment:
Really not sure about this one: The mix of private declaration and public alias (to a pointer of it) smells bad. Do we reallty need a subclass for this? If so, why not make it a standallone class -- or fully move it into the implementation file and just use a declaration here?

Reply:
I kept the declaration, since the distinction to regular messages is important, but moved the details to the implementation, as suggested.